include/yaclib/async/detail/core.hpp

```diff
@@ -48,32 +44,24 @@ class BaseCore : public CallerCore {
       Execute();
     }
   }
-
-  executor::IExecutorPtr GetExecutor() const noexcept {
-    return _executor;
+  void Stop() noexcept {
+    _state.store(State::Stopped, std::memory_order_release);
+    assert(_callback == nullptr);
   }
 
-  bool SetWaitCallback(util::Ptr<ITask> callback) {
+  bool SetWaitCallback(util::Ptr<ITask> callback) noexcept {
     _callback = std::move(callback);
-    const auto state = _state.exchange(State::HasCallback, std::memory_order_acq_rel);
-    const bool ready = state == State::HasResult;  // this is mean we have result
+    const auto state = _state.exchange(State::HasWaitCallback, std::memory_order_acq_rel);
+    const bool ready = state == State::HasResult;  // This is mean we have Result
     if (ready) {
-      _callback = nullptr;
       _state.store(State::HasResult, std::memory_order_release);
+      _callback = nullptr;
```

Comment:
```suggestion
      _callback.Release();
```
Here we do a useless `mutex.lock() and `cv.notify()` when we can just forget about this callback.

Reply:
You are right, but the suggestion code won't work for WaitGroup